Npm7 fix

bizob2828 opened the pull request:
<!--
Thank you for submitting a Pull Request.

This code is leveraged to monitor critical services. Please consider the following:
* Tests are required.
* Performance matters.
* Features that are specific to just your app are unlikely to make it in.

Please fill out the relevant sections as follows:
* Proposed Release Notes: Bulleted list of recommended release notes for the change(s).
* Links: Any relevant links for the change.
* Details: In-depth description of changes, other technical notes, etc.
-->

## Proposed Release Notes
 * Provides ability to install every package on each test run

## Details
npm 7 breaks how our versioned test runner works.  In npm pre 7 when you run `npm install <pkg> --no-save --no-package-lock` it would only install the packages you specified and leave any unrelated to its dep tree untouched.  With npm 7 it cleans up any packages not related to installing said packages.  After much thought I think this was always a bug us Node.js devs accepted.  Our use case of using `--no-save --no-package-lock` is very rare.  I did provide this as a flag `-a`.  This means that when we run our v16 versioned tests we will have to provide the `-a` flag.  I did some test runs with this flag on and pre npm 7 it's slower but post npm 7 it's about the same.

## Timing

Before fix on agent(npm v6)
```
PASS

real	1m32.308s
user	3m31.095s
sys	0m53.817s

real	1m32.340s
user	3m31.105s
sys	0m53.831s
```

After fix on agent(npm v6)
```
real	2m32.878s
user	7m0.625s
sys	2m0.400s

real	2m32.934s
user	7m0.633s
sys	2m0.415s
```

After fix on agent(npm v7)
```
real	1m25.573s
user	3m46.563s
sys	0m59.746s

real	1m25.630s
user	3m46.571s
sys	0m59.762s
```


Before fix on aws-sdk(npm v6)
```
real	1m42.060s
user	1m28.913s
sys	0m32.542s
```

After fix on aws-sdk(npm v6)
```
real	3m51.166s
user	3m9.770s
sys	1m33.439s
```

After fix on aws-sdk(npm v7)
```
real	1m12.032s
user	1m00.820s
sys	0m24.402s
```

bizob2828 commented on bin/version-manager.js:
```diff
@@ -28,6 +28,7 @@ cmd
   .option('--major', 'Only iterate on major versions of packages.')
   .option('--minor', 'Iterate over minor versions of packages (default).')
   .option('--patch', 'Iterate over every patch version of packages.')
+  .option('-a, --all', 'Installs all packages, not just ones that differ in version')
```
I'd like to get feedback on this because by default this will only install packages that differ from previous run.  But when we wire up node 16 versioned tests we want this flag to be set to it always installs every package, but since it's a CLI option and we wrap this in a npm script it might be hard to do so

carlo-808 replied:
We can set the CI to run the version tests with this flag on for Node 16 only when we get to that work. Looks like we would have to pipe in the flag to this [point](https://github.com/newrelic/node-newrelic/blob/main/bin/run-versioned-tests.sh#L24) through a package.json script adding a commandline option to the `run-versioned-tests.sh` script in our agent. Seems ok to me.

michaelgoin replied:
its easy enough to pass options through the npm scripts we use. seems like a good start. if we find the perf hit isn't too bad (or not noticeable cause npm 7 is faster anyways) then we can prob flip the default.

michaelgoin reviewed (approved):
👍

bizob2828 commented on lib/versioned/runner.js:
```diff
@@ -71,7 +71,14 @@ function installPackages(cb) {
+      '--no-package-lock', // do not update package-lock file
+      '--no-audit', // skip audit output
+      '--no-fund', // skip funding output
+      '--legacy-peer-deps' // skip installing peer deps
```
i need to look into the test run from the agent because I think we do need this option

bizob2828 replied:
Looking into this test it might be wrong?

the deps are
```
    "dependencies": {
        "ejs": "2.5.5",
        "hapi": ">=17.0.0",
        "vision": "5.1"
      },
```

but the folder implies this should only be hapi 17 but it is also running against hapi 18.  This fails with npm 7 because vision 5.1 has a peer dep of  https://github.com/hapijs/vision/blob/v5.1.0/package.json#L22. should I open a separate PR to lock down hapi to `>=17.0.0 < 18.0.0`

bizob2828 commented:
It's also worth noting we need to fix our cls versioned tests.  npm 7 doesn't install optional deps that cannot be satisfied. v2.6.2 has an optional dep of `async-listener: 0.4.5` https://github.com/othiym23/node-continuation-local-storage/blob/v2.6.2/package.json#L34.  This version only allows node 0.11.8 or less https://github.com/othiym23/async-listener/blob/v0.4.5/package.json#L30

```
/Users/revans/code/pull-requests/node-newrelic/test/versioned/cls/cls.tap.js continuation-local-storage@2.6.2
stdout

added 2 packages, and changed 1 package in 788ms
TAP version 13
# Subtest: transaction state
    1..1
    # Subtest: should be preserved over timers regardless of order required
        not ok 1 - Cannot find module 'async-listener' Require stack: - /Users/revans/code/pull-requests/node-newrelic/test/versioned/cls/node_modules/continuation-local-storage/context.js - /Users/revans/code/pull-requests/node-newrelic/test/versioned/cls/cls.tap.js
          ---
          stack: >
            Object.<anonymous> (node_modules/continuation-local-storage/context.js:15:32)

            Module._compile (/Users/revans/code/pull-requests/node-newrelic/node_modules/source-map-support/source-map-support.js:547:25)
          at:
            line: 880
            column: 15
            file: internal/modules/cjs/loader.js
            function: Function.Module._resolveFilename
          code: MODULE_NOT_FOUND
          requireStack:
            - /Users/revans/code/pull-requests/node-newrelic/test/versioned/cls/node_modules/continuation-local-storage/context.js
            - /Users/revans/code/pull-requests/node-newrelic/test/versioned/cls/cls.tap.js
          tapCaught: testFunctionThrow
          test: should be preserved over timers regardless of order required
          ...

        1..1
        # failed 1 test
    not ok 1 - should be preserved over timers regardless of order required # time=27.732ms

    # failed 1 test
not ok 1 - transaction state # time=48.372ms

1..1
# failed 1 test
# time=71.719ms
```
